### PR TITLE
Support a read-only option to vsphere volumes to mount volumes as read-only on container run.

### PR DESCRIPTION
**Pull request by govint**
Added an option to specify an access mode on volume creation, the volume is mounted always as read-only when attached to a container. Which is ok only as long as there is a way to attach the volume to the VM and be able to init. the volume with data. Perfect for pre-initialized volumes being imported into the volume repo on the vsphere side (issue #442 ). 

Tested as below:

1. Create a volume with access=read-only and mount it. Verify it mounts read-only on the client.

administrator@hte-1s-eng-dhcp98:~/vpl-add-ro-vols/docker-volume-vsphere$ docker volume inspect testvol-4
[
    {
        "Name": "testvol-4",
        "Driver": "vmdk",
        "Mountpoint": "/mnt/vmdk/testvol-4",
        "Status": {
            "access": "read-only",
            "capacity": {
                "allocated": "14.0MB",
                "size": "100.0MB"
            },
            "created": "Mon Aug 22 07:27:35 2016",
            "created by VM": "2",
            "datastore": "bigone",
            "status": "detached"
        },
        "Labels": {},
        "Scope": "global"
    }
]
administrator@hte-1s-eng-dhcp98:~/vpl-add-ro-vols/docker-volume-vsphere$ docker run --rm -it -v testvol-4:/data busybox
/ # cd /data
/data # >sdafhwef
sh: can't create sdafhwef: Read-only file system
/data #

And /proc/mounts shows the read-only mount,
/dev/disk/by-path/pci-0000:13:00.0-scsi-0:0:0:0 /mnt/vmdk/testvol-4 ext2 ro,relatime 0 0

2. Try using an invalid access type:

$ docker volume create --driver=vmdk --name=testvol-5 -o "access=access1"
Error response from daemon: create testvol-5: VolumeDriver.Create: Access type 'access1' is not supported. Valid options are: ['read=write', 'read-only']

3. Verify default behavior is to create a read-write volume

$ docker volume create -d vmdk --name testvol-6
testvol-6

$ docker volume inspect testvol-6
[
    {
        "Name": "testvol-6",
        "Driver": "vmdk",
        "Mountpoint": "/mnt/vmdk/testvol-6",
        "Status": {
            "access": "read-write",
            "capacity": {
                "allocated": "14.0MB",
                "size": "100.0MB"
            },
            "created": "Mon Aug 22 11:51:26 2016",
            "created by VM": "2",
            "datastore": "bigone",
            "status": "detached"
        },
        "Labels": {},
        "Scope": "global"
    }
]

### CODE REVIEW

**Comment by msterin** [comment]
Fixes #274

**Comment by pdhamdhere** [comment]
I think I am missing something. Help me understand the proposed feature;
-  If volume is created as read-only then how does user initializes it?
- Can this volume be attached to multiple docker hosts as read-only?
- Would it support 1-writer (Docker Host) and many-readers (Docker hosts)?

**Comment by msterin** [comment]
@pdhamdhere , let me taker a stub at answering per my understanding:

    If volume is created as read-only then how does user initializes it

since "read only" behavior is controlled by KV, nothing prevents us from adding an admin CLI option to flip to on and off for offline volumes. With this, the question "how do you create it" is answered.  So the PR needs the admin support too.

    Can this volume be attached to multiple docker hosts as read-only?

I think mounting to multiple dockers still won't work as VMX/VMFS interaction will still lock the vmdk and treat it as writeable. We can craft something around delta disks, but as-is this is still exclusive use.

    Would it support 1-writer (Docker Host) and many-readers (Docker hosts)?

No

**Comment by msterin** [comment]
FWIW I think this PR, while the code is good,  requires Admin CLI support and test automation to be functional.

**Comment by govint** [comment]
1. The admin CLI command is this one:
python /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py edit -h
usage: vmdkops_admin.py edit [-h] --volume VOLUME --options OPTIONS

optional arguments:
  -h, --help         show this help message and exit
  --volume VOLUME    Full path of the volume
  --options OPTIONS  Options (specifically, access) to be set on the volume.

options is a comma separated list jut like how options are specified when creating a volume.

2. Using the command to reset access mode on a volume

$ docker volume create -d vmdk --name testvol-11
testvol-11

$ docker volume inspect testvol-11
[
    {
        "Name": "testvol-11",
        "Driver": "vmdk",
        "Mountpoint": "/mnt/vmdk/testvol-11",
        "Status": {
            "access": "read-write",
            "attach_as": "independent_persistent",
            "capacity": {
                "allocated": "14.0MB",
                "size": "100.0MB"
            },
            "created": "Thu Aug 25 08:05:33 2016",
            "created by VM": "2",
            "datastore": "bigone",
            "diskformat": "thin",
            "status": "detached"
        },
        "Labels": {},
        "Scope": "global"
    }
]

On the host reset the access mode on the volume

] python /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py edit --volume=/vmfs/volumes/bigone/dockvols/testvol-11.vmdk --opti
ons="access=read-write"
Successfully updated settings for : /vmfs/volumes/bigone/dockvols/testvol-11.vmdk

$ docker volume inspect testvol-11
[
    {
        "Name": "testvol-11",
        "Driver": "vmdk",
        "Mountpoint": "/mnt/vmdk/testvol-11",
        "Status": {
            "access": "read-write",
            "attach_as": "independent_persistent",
            "capacity": {
                "allocated": "14.0MB",
                "size": "100.0MB"
            },
            "created": "Thu Aug 25 08:05:33 2016",
            "created by VM": "2",
            "datastore": "bigone",
            "diskformat": "thin",
            "status": "detached"
        },
        "Labels": {},
        "Scope": "global"
    }
]

File: esx_service/cli/vmdkops_admin.py [hunk 1]
```diff
@@ -255,6 +255,20 @@ def commands():
         'status': {
             'func': status,
             'help': 'Show the status of the vmdk_ops service'
+        },
+        'edit': {
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: 'edit' sounds like we are editing content. 'settings' or 'config' sounds better to me . `config --options`

File: esx_service/cli/vmdkops_admin.py [hunk 2]
```diff
@@ -515,6 +529,14 @@ def status(args):
     print("LogLevel: {0}".format(log_config.get_log_level()))
 
 
+def edit_vol(args):
+    edit_ok = vmdk_ops.edit_vol_opts(args.volume, args.options) 
+    if edit_ok:
+        print('Successfully updated settings for : {0}'.format(args.volume))
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: printing message here seems unneeded.  Do we do it in other places ? I think we do need a log record though .

File: esx_service/cli/vmdkops_admin.py [hunk 3]
```diff
@@ -515,6 +529,14 @@ def status(args):
...
+    else:
+        print('Edit failed for {0}.'.format(args.volume))
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: will it tell why it failed? An example of output would be good.

File: esx_service/vmdk_ops.py [hunk 4]
```diff
@@ -996,6 +1019,33 @@ def disk_detach_int(vmdk_path, vm, device):
     return None
 
 
+# Edit settings for a volume identified by its full path
+def edit_vol_opts(vmdk_path, options):
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: nit: an example of parsed input, in comments, would be good

**Comment by msterin** [comment]
Nice change !

Comments: 'edit' in admin options sounds like we are editing the volume (not config) so I suggest 'set'. Also, I do not think we need a message on success (unless we do it in all other admin commands', and instead of that we need a log record.  A few more nits inline. 

LGTM after that.

File: esx_service/cli/vmdkops_admin.py (hunk 1 above)
Comment (andrewjstone) [inline esx_service/cli/vmdkops_admin.py]: I agree. `update` is also a possibility. `set` is used in the role commands, while `update` is used in the policy command.

**Comment by andrewjstone** [comment]
Nice work @govint LGTM after changing 'edit' to 'set'.`

**Comment by pdhamdhere** [comment]
Please update documentation with new options in same PR. /cc @kerneltime

**Comment by msterin** [comment]
LGTM

File: docs/misc/vmdkops-admin-cli-spec.md [hunk 5]
```diff
@@ -121,6 +121,18 @@ Show any interesting information about the service. This includes file paths of
...
 that the exact format may be somewhat different.
 
+# Set
+Modify attribute settings on a given volume. The volume is identified by its full path on the datastore,
+for example if the volume name is `container-vol` then the volume is specified as "/vmfs/volumes/<datastore>/dockvols/container-vol.vmdk".
+The attributes to set/modify are specified as a comma separated list as "<attr1>=<value>, <attr2>=<value>....". For example,
+a command line would look like this.
+
+```
+$ vmdkops-admin set --volume=<full path of volume> --options="<attr1>=<value>, <attr2>=<value>, ..."
```
Comment (pdhamdhere) [inline docs/misc/vmdkops-admin-cli-spec.md]: Should we be using volumeName@datastore similar to what we use in docker run command. I don't think vmdkops-admin prints full volume path in any command.

File: docs/user-guide/docker-volume-cli.md [hunk 6]
```diff
@@ -54,3 +54,11 @@ Currently the following are supported
...
 2. [independent_persistent](http://cormachogan.com/2013/04/16/what-are-dependent-independent-disks-persistent-and-non-persisent-modes/): If the VMDK is attached as independent_persistent it will not be part of a VM snapshot. The Docker volume can be attached to any VM that can access the datastore independent of snapshots.
+
+### access
+```
+docker volume create --driver=vmdk --name=MyVolume -o access=read-only -o diskformat=thin
+docker volume create --driver=vmdk --name=MyVolume -o access=read-write -o diskformat=thin (default)
+```
+
+The access mode determines if the volume is modifiable by containers in a VM. The access mode allows to first create a volume with write access and initialize it with binary images, libraries (for exmple), and subsequently change the access to "read-only" (via the admin CLI). Thereby, creating content sharable by all containers in a VM.
```
Comment (pdhamdhere) [inline docs/user-guide/docker-volume-cli.md]: Add "Volume must not be in use"?

Comment (pdhamdhere) [inline docs/user-guide/docker-volume-cli.md]: BTW, I don't see a check for whether VMDK is attached?

Comment (govint) [inline docs/user-guide/docker-volume-cli.md]: The command only updates the KV and will take effect on the next attach
I'll update that in the doc.

File: docs/misc/vmdkops-admin-cli-spec.md (hunk 5 above)
Comment (govint) [inline docs/misc/vmdkops-admin-cli-spec.md]: The issue is the plugin gets to know the VM and can figure which dockvols
folder to create the volume in. But the user on the ESX always needs to
know the datastore where the volume is located. Otherwise we can't locate
the volume at all. Hence the need to give the path.

Alternative is for the user to always give volume@datastore as the format.
I can change to make it that way.

**Comment by kerneltime** [comment]
If this is still a single VM attach then I don't see how it fixes https://github.com/vmware/docker-volume-vsphere/issues/274 what feature are we implementing here? Also, we need to add tests, documentation and demos.

**Comment by govint** [comment]
Both documentation and tests are included in the change. It's not possible
to attach a volume to multiple VMs without making it a shared volume in
vmfs.

File: docs/user-guide/docker-volume-cli.md (hunk 6 above)
Comment (govint) [inline docs/user-guide/docker-volume-cli.md]: Volume can be in use. Only the kV is being changed and will l take effect
on the next mount inside the guest which can happen after a detach and
subsequent attach

File: docs/misc/vmdkops-admin-cli-spec.md (hunk 5 above)
Comment (pdhamdhere) [inline docs/misc/vmdkops-admin-cli-spec.md]: I think ' vmdkops-admin ls ' prints datastore. VolName@DS in addition to full path would be nice and can be separate PR.

File: esx_service/cli/vmdkops_admin.py [hunk 7]
```diff
@@ -255,6 +255,20 @@ def commands():
         'status': {
             'func': status,
             'help': 'Show the status of the vmdk_ops service'
+        },
+        'set': {
+            'func': set_vol_opts,
+            'help': 'Edit settings for a given volume',
+            'args': {
```
Comment (pdhamdhere) [inline esx_service/cli/vmdkops_admin.py]: Sure. Add a note in comment that change will be effective next time volume is mounted or something on those lines.

**Comment by pdhamdhere** [comment]
LGTM after updating help

**Comment by kerneltime** [comment]
fixes https://github.com/vmware/docker-volume-vsphere/issues/413

File: docs/misc/vmdkops-admin-cli-spec.md (hunk 5 above)
Comment (kerneltime) [inline docs/misc/vmdkops-admin-cli-spec.md]: This is the wrong place to update the user facing CLI documentation.. fixing it and posting a PR..
